### ISSUE TITLE
Set ownerReference on gateway Deployment

### ISSUE THREAD

**Issue by JuniorJPDJ** [envoyproxy/gateway#3051]
*Description*:
Would be cool to see `Deployment` created for `Gateway` to have `ownerReference` for `Gateway`.
This allows following chain and for example makes ArgoCD show the deployment where `Gateway` is created in ArgoCD application.

**Comment by cnvergence** [comment]
+1

**Comment by arkodg** [comment]
rethinking, this work may be a little bit more involved, we lose exact info of the `Kind` and `Namespace` & `Name` when converting from gateway-api to IR, so this will cost some extra memory
https://github.com/envoyproxy/gateway/blob/3d519339bc68ab1f8ece22c89aeb757e7fcc58cd/internal/ir/infra.go#L39

**Comment by zirain** [comment]
same to `service`? and what about mergedGateway?

**Comment by cnvergence** [comment]
merged gatewaya should ref the GatewayClass

**Comment by arkodg** [comment]
not the prettiest and slightly fragile, but the easiest solution is to update the ownerReference based on the owning labels created https://github.com/envoyproxy/gateway/blob/8a01dd68ae8c352c438292fdb93316d9badd35d8/internal/gatewayapi/helpers.go#L230

**Comment by arkodg** [comment]
this cannot be implemented with the default deployment configuration because ownerReferences dont have cross namespace support, the `Gateway` exists in `ns1` and the generated resources are deployed in the `envoy-gateway-system` by default

**Comment by arkodg** [comment]
this can only work for the mergeGateways case where the owner is a cluster scoped resource like `GatewayClass`

**Comment by sadovnikov** [comment]
> not the prettiest and slightly fragile, but the easiest solution is to update the ownerReference based on the owning labels created

@arkodg, I tried adding the `metadata.ownerReferences` manually. Argo CD picks up and shows the Deployment and the Service. However, the labels are not enough because `uid` of the owner is a required attribute

```
A copy of your changes has been stored to "/tmp/kubectl-edit-559773258.yaml"
error: map: map[apiVersion:gateway.networking.k8s.io/v1 blockOwnerDeletion:false kind:Gateway name:default] does not contain declared merge key: uid
```

![Image](https://github.com/user-attachments/assets/c6e40dc4-51c1-4dff-91fc-d90a57801abc)

**Comment by arkodg** [comment]
@sadovnikov one way to specifically solve ^ is to set a label in `Gateway.Spec.Infrastructure.Labels`

**Comment by kkk777-7** [comment]
Hello @arkodg, I understand that the following PR allows creating a Deployment/Service in the user namespace.
https://github.com/envoyproxy/gateway/pull/5137

In this case, I believe it addresses the issue you mentioned in [comment](https://github.com/envoyproxy/gateway/issues/3051#issuecomment-2591281804).
how about setting an ownerReference when using GatewayNamespace mode?

I think users who want resources to be automatically cleaned up can use GatewayNamespace mode. (should consider how the migration would be handled.)

If this approach is acceptable, I'd be happy to take on the implementation work. WDYT?

**Comment by arkodg** [comment]
also for the default mode (`ControllerNamespace`) should we set the ownerReference to the `envoy-gateway` `Deployment` ?

**Comment by kkk777-7** [comment]
@arkodg thanks for reply, i'll proceed with the implementation for the GatewayNamespace case.


> also for the default mode (ControllerNamespace) should we set the ownerReference to the envoy-gateway Deployment ?

IMO, since there's a resource relationship between Gateway and Deployment/Service, Gateway should be owner if we use ownerReference.
But, as you pointed out, it's annoying that can't be achieved due to the limitations of ownerReferences(cross namespace support).

**Comment by arkodg** [comment]
keeping this open to make `GatewayClass` an owner reference for the default case

**Comment by kkk777-7** [comment]
@arkodg Sure, I'll get on it! I have one question - does this apply to all cases, or only in the case of merged gateway?

**Comment by arkodg** [comment]
i'll vote for all cases, wdyt @envoyproxy/gateway-maintainers ?

The reason for this is having a grand parent (GatewayClass) is better than not having anything for https://github.com/envoyproxy/gateway/issues/3051#issuecomment-2710770054

will need to dig into the downside of making a grandparent a owner

**Comment by kkk777-7** [comment]
@arkodg @cnvergence 
thanks for comments. I will implement it for all cases.

**Comment by zhaohuabing** [comment]
Sorry for jumping in late: If we set the OwnerReference to GatewayClass instead of Gateway, does that mean [Kubernetes won‘t be able to clean up the leftover resources after a Gateway is deleted](https://github.com/envoyproxy/gateway/pull/6173#discussion_r2125590501)? @arkodg

**Comment by arkodg** [comment]
@zhaohuabing we can't set ownerReference to the Gateway for the default case due to https://github.com/envoyproxy/gateway/issues/3051#issuecomment-2591281804

**Comment by zhaohuabing** [comment]
> [@zhaohuabing](https://github.com/zhaohuabing) we can't set ownerReference to the Gateway for the default case due to [#3051 (comment)](https://github.com/envoyproxy/gateway/issues/3051#issuecomment-2591281804)

How about GatewayNamespace mode?  Do we use the same strategy(always pointing to GatewayClass) to simplify the ownerreference logic?

**Comment by arkodg** [comment]
Yes for GNM it's set to Gateway https://github.com/envoyproxy/gateway/pull/6100
